src/kademlia/dht_storage.cpp

```diff
@@ -0,0 +1,532 @@
+		if (i == m_mutable_table.end())
+		{
+			// this is the case where we don't have an item in this slot
+			// make sure we don't add too many items
+			if (int(m_mutable_table.size()) >= m_settings.max_dht_items)
+			{
+				// delete the least important one (i.e. the one
+				// the fewest peers are announcing)
+				dht_mutable_table_t::iterator j = std::min_element(m_mutable_table.begin()
+					, m_mutable_table.end()
+					, boost::bind(&dht_immutable_item::num_announcers
+						, boost::bind(&dht_mutable_table_t::value_type::second, _1)));
+				TORRENT_ASSERT(j != m_mutable_table.end());
+				free(j->second.value);
+				free(j->second.salt);
+				m_mutable_table.erase(j);
+				m_counters.inc_stats_counter(counters::dht_mutable_data, -1);
+			}
+			dht_mutable_item to_add;
+			to_add.value = static_cast<char*>(malloc(size));
+			to_add.size = size;
+			to_add.seq = seq;
+			to_add.salt = NULL;
+			to_add.salt_size = 0;
+			if (salt_size > 0)
+			{
+				to_add.salt = static_cast<char*>(malloc(salt_size));
+				to_add.salt_size = salt_size;
+				memcpy(to_add.salt, salt, salt_size);
+			}
+			memcpy(to_add.sig, sig, sizeof(to_add.sig));
+			memcpy(to_add.value, buf, size);
+			memcpy(&to_add.key, pk, sizeof(to_add.key));
+
+			boost::tie(i, boost::tuples::ignore) = m_mutable_table.insert(
+					std::make_pair(target, to_add));
+			m_counters.inc_stats_counter(counters::dht_mutable_data);
+
+//			fprintf(stderr, "added mutable item (%d)\n", int(m_mutable_table.size()));
+		}
+		else
+		{
+			// this is the case where we already
+			dht_mutable_item* item = &i->second;
+
+			if (item->seq < seq)
+			{
+				if (item->size != size)
+				{
+					free(item->value);
+					item->value = static_cast<char*>(malloc(size));
+					item->size = size;
+				}
+				item->seq = seq;
+				memcpy(item->sig, sig, sizeof(item->sig));
+				memcpy(item->value, buf, size);
+			}
+		}
+
+		touch_item(&i->second, addr);
+	}
+
+	void dht_default_storage::tick()
+	{
+		time_point now(aux::time_now());
+
+		for (dht_immutable_table_t::iterator i = m_immutable_table.begin();
+			 i != m_immutable_table.end();)
+		{
+			if (i->second.last_seen + minutes(60) > now)
+			{
+				++i;
+				continue;
+			}
+			free(i->second.value);
+			m_immutable_table.erase(i++);
+			m_counters.inc_stats_counter(counters::dht_immutable_data, -1);
+		}
+
+		// look through all peers and see if any have timed out
+		for (table_t::iterator i = m_map.begin(), end(m_map.end()); i != end;)
+		{
+			torrent_entry& t = i->second;
+			node_id const& key = i->first;
+			++i;
+			purge_peers(t.peers);
+
+			if (!t.peers.empty()) continue;
+
+			// if there are no more peers, remove the entry altogether
+			table_t::iterator it = m_map.find(key);
+			if (it != m_map.end())
+			{
+				m_map.erase(it);
+				m_counters.inc_stats_counter(counters::dht_torrents, -1);
+			}
+		}
+	}
+
+	dht_storage_interface* dht_default_storage_constructor(sha1_hash const& id
```

Comment:
I suppose boost::scoped_ptr can't be copied, so we have to resort to a raw pointer here to stay c++98 compatible.